### ISSUE TITLE
Use Case Notebook for "Atmospheric Moisture Budgets"

### ISSUE THREAD

**Issue by rabernat** [pangeo-data/pangeo-discussion#1]
We need a preliminary notebook that uses Xarray to calculate something related to this use case. It doesn't have to be super long or complicated. It just has to express a real-world, scientifically relevant calculation on a large dataset.

Requirements:

- [ ] Carefully explain the science behind the calculation in words and equations.
- [ ] Clearly document the input dataset (preferably a standard product like ERA-interim or NCAR large ensemble)
- [ ] Use Xarray to load the dataset and, as much as possible, to make the calculations
- [ ] Where Xarray can't accomplish the necessary tasks, clearly document the other functions and / or outside packages that need to be employed
- [ ] If the analysis has been implemented in other languages (e.g. MATLAB, Ingrid, etc.), provide a link to the code and, if possible, the output (which can be used as validation)

cc: @naomi-henderson

I would like to use this as a template for the other three use cases, so I would appreciate general feedback on the checklist above. Are these the right requirements?

**Comment by rabernat** [comment]
We should add something about setting up the environment and enumerating the dependencies.

Here is my guide to setting up a conda environment for xarray work:
https://medium.com/@rabernat/custom-conda-environments-for-data-science-on-hpc-clusters-32d58c63aa95

But @darothen's "getting started" guide for gcpy is better and more comprehensive:
http://www.danielrothenberg.com/gcpy/getting_started.html

**Comment by rabernat** [comment]
You also asked about the gist extension:
https://github.com/mozilla/jupyter-notebook-gist

**Comment by darothen** [comment]
Please feel free to cannibalize my guide. If I can help in any way, do let me know.

**Comment by spencerahill** [comment]
Curious, what motivated this (and "the other three") use case?  Just something that satisfies your above constraints?

I ask because I have worked on the problem of closing atmospheric column tracer budgets computed post-hoc.  Depending on how concerned you are with closure, this can be pretty involved.  The Appendices of our recent paper touch on this: http://journals.ametsoc.org/doi/abs/10.1175/JCLI-D-16-0785.1.  

But I suspect this is more detail than you are looking for?  If not, I can provide the Python functions that I defined for these efforts (they relied on my [infinite-diff](https://github.com/spencerahill/infinite-diff) package and were computed via an [aospy](https://github.com/spencerahill/aospy) pipeline).

**Comment by rabernat** [comment]
@spencerahill: hopefully the [project description](https://figshare.com/articles/Pangeo_NSF_Earthcube_Proposal/5361094) I published earlier today clarifies what I meant by "use cases": these are the official use cases from the NSF proposal. However, we are certainly in need of as many use cases as possible and welcome contributions from everyone. If you have use cases you want to contribute, I would submit a PR to the [pangeo-tutorials](https://github.com/pangeo-data/pangeo-tutorials) repo (currently empty, but hopefully will fill up with cool stuff!)

**Comment by rabernat** [comment]
Just a heads-up that the lack of any use case notebooks is now blocking progress on our project. It would be great to get some basic use cases checked in. They don't have to be fancy! Any sort of semi-realistic workflow on real data is enough for the systems people to move forward with analyzing performance.

**Comment by naomi-henderson** [comment]
The calculation of the eddy moisture transports is the most intensive I/O part of the budgets, so I am developing a preliminary notebook which computes the monthly mean co-variances from 6 hourly (e.g., u'q').  I will get the notebook working on cheyenne for the CMIP5 version of CESM.  Once it is working, I will submit it to the group for critique before cleaning it up.  Then I will start working on the vertical integrals, etc.  Sound ok?

**Comment by mrocklin** [comment]
@naomi-henderson if you run into performance issues (like I/O bottlenecks) and want to try scaling things let me know.  I'd be happy to try to collaborate to try to accelerate things.

**Comment by rabernat** [comment]
@naomi-henderson your plan sounds perfect

Also feel free to ping us for help with any xarray or general python questions you may encounter while developing your notebook. I know you're still spinning up on these tools, so we are here to help with that too.

**Comment by naomi-henderson** [comment]
@rabernat  - I have made a preliminary version of an eddy covariance notebook and it works fine, albeit slowly, on my local machine (one year has to process 24G).  I am having a few issues with the time grids, groupby, etc, as you can see here: [covar notebook](https://github.com/naomi-henderson/pangeoMB/blob/master/covar-SingleYear.ipynb)   I may have wandered off into the deep end ...

**Comment by rabernat** [comment]
Awesome @naomi-henderson! This looks like a great start. I'm very pleased you have gotten your calculation to a preliminary working version.

I would love to sit with you for an hour on Wednesday and go though this in detail. It should be possible to eliminate the loops over files. Dask is unable to parallelize such operations, and they can also be confusing to read. The idea is to first create one big xarray dataset with everything in it (although not in memory) and then perform the desired grouping / aggregation on it using xarray's built in capabilities. You should never have to manually accumulate an average, as you do in your notebook--although this is a very natural place to start for someone coming from matlab or another lower level language!

I'm free all day on Wednesday, so please suggest a time that works for you!

**Comment by mrocklin** [comment]
It might be useful to keep this notebook around to show before-and-after examples.  This Rosetta-stone-style comparison might help other scientists more accustomed to C++/Fortran style development understand XArray-style development and vice versa.

**Comment by mrocklin** [comment]
We only have a few opportunities for this kind of comparison before everyone gets exposed to both types.

**Comment by naomi-henderson** [comment]
Thanks @rabernat, I was hoping you would have time to help me with this!  Does 10am on Wed work for you?

**Comment by rabernat** [comment]
Yes, perfect.

**Comment by naomi-henderson** [comment]
Thanks to @rabernat this morning, I have a new version of the preliminary eddy covariance notebook:  [New version, no loops](https://github.com/naomi-henderson/pangeoMB/blob/master/covar-SingleYear-ver2.0.ipynb). I love the suggestion by @mrocklin to keep my preliminary notebooks around to help others migrate to the xarray/dask world!

I hope to check it out on cheyenne (using the 'export LD_LIBRARY_PATH=' temporary fix) and
perhaps even using the dask dashboard at some point.

My next step is to make a new notebook which does the vertical interpolation of the monthly data from sigma levels to pressure levels.  I know that there are python scripts to do this on the command line file by file, but we need something to work on the whole xarray Dataset. Expect many questions.

**Comment by mrocklin** [comment]
Cool looking notebook @naomi-henderson !  Makes me wish I understood what you all were doing :)

Do you have any thoughts about the experience?  Both on what felt pleasant and on what felt frustrating at first?  It would be useful to understand your thought process as you're going through this.

**Comment by mrocklin** [comment]
(only if you feel like sharing, didn't mean to put you on the spot)

**Comment by naomi-henderson** [comment]
I forgot to mention the (many) reasons why the eddy covariance notebook is still preliminary.  Although it works for the NCAR/CCSM4 model, it will not work for the models with multiple ensemble members. This will require adding a new 'member' dimension.  

The notebook also needs various checks for the existence of the correct variables as well as debugging for all possible calendars (leap, no leap, 360 days/year, etc).

**Comment by rabernat** [comment]
> Makes me wish I understood what you all were doing :)

@mrocklin: the markdown cell at the end of the notebooks gives a description of the calculation and data used. But it is probably still assumes too much domain knowledge and uses lots of jargon. Let us know how we can improve this to make it more understandable to a wider audience.

**Comment by rabernat** [comment]
@spencerahill

> I forgot to mention the (many) reasons why the eddy covariance notebook is still preliminary. Although it works for the NCAR/CCSM4 model, it will not work for the models with multiple ensemble members. This will require adding a new 'member' dimension.
> 
> The notebook also needs various checks for the existence of the correct variables as well as debugging for all possible calendars (leap, no leap, 360 days/year, etc).

@spencerahill: any thoughts on how [aospy](http://aospy.readthedocs.io/en/latest/) might fit into the CMIP data processing workflow that Naomi is describing above?

**Comment by spencerahill** [comment]
>Although it works for the NCAR/CCSM4 model, it will not work for the models with multiple ensemble members. This will require adding a new 'member' dimension.

>The notebook also needs various checks for the existence of the correct variables as well as debugging for all possible calendars (leap, no leap, 360 days/year, etc).

>any thoughts on how aospy might fit into the CMIP data processing workflow that Naomi is describing above?

These kind of things are definitely in the spirit of aospy, but it may be overkill to implement a full aospy-based workflow in this case.  Frankly, we require a fair amount of boilerplate at present, which I think cuts against the intent of the notebook.  (Perhaps, @rabernat, a separate notebook specifically presenting an aospy workflow for automating calculations across CMIP runs would be welcome?)

For ensemble members, this is where xarray's data model shines: insert a new dimension into each dataset called e.g. `ens_mem`, which then they get combined over via [`xr.concat`](http://xarray.pydata.org/en/stable/generated/xarray.concat.html).  It could be without a coordinate, in which case it simply indexes the different ensemble members.  Or, you could add an `ens_mem` coordinate as well that stores the ensemble member's string identifier, which for CMIP data looks like `'r1i1p1'`,`'r2i1p1'`, etc.  Also, aospy's support for ensemble members is currently broken: see https://github.com/spencerahill/aospy/issues/76 😆 

For the calendars, again xarray does the heavy lifting -- so long as the calendars and time data are CF-compliant (which CMIP data all should be), xarray should decode it properly without much trouble.  Where aospy adds additional functionality is in dealing with dates outside the (dare I say narrow) range of permissible dates for `np.datetime64` (@naomi-henderson, see e.g. [here](https://github.com/pydata/xarray/issues/1084) for background).  This is not an issue for the example given currently in the notebook, but it is in many other cases.

See e.g. [aospy.utils.times.numpy_datetime_workaround_encode_cf](https://aospy.readthedocs.io/en/latest/api.html#aospy.utils.times.numpy_datetime_workaround_encode_cf), which re-encodes the time array of a Dataset with out-of-bounds dates that has not been CF-decoded such that, when it is decoded, the time array starts at (nearly) the earliest permissible date.  @spencerkclark has much more expertise on calendars/datetimes than me, both within aospy and without, so Spencer feel free to chime in.

**Comment by naomi-henderson** [comment]
Thanks @spencerahill for the hints on creating a new dimension for ensemble member.  

I have now tried the [eddy covariance notebook](https://github.com/naomi-henderson/pangeoMB/blob/master/covar-SingleYear-ver2.0.ipynb) on cheyenne, using Jupyter Lab and the Dask dashboard.  @mrocklin  and @rabernat , I am certainly having performance issues, but it looks like it is mostly reading in the CMIP5 data from '/glade/p/CMIP/CMIP5/output1/...'

The open_mfdataset call  
``` %time ds = xr.open_mfdataset(all_files, preprocess=set_correct_coords) ```
takes over 5 min on cheyenne: 
```
CPU times: user 4.93 s, sys: 3.17 s, total: 8.1 s
Wall time: 5min 13s
```
Whereas the same on my linux box here at Lamont (data stored on internal 4Tb drive)  takes less than 10s:
```
CPU times: user 5.91 s, sys: 5.51 s, total: 11.4 s
Wall time: 8.31 s
```

However the resampling and load() from 6-hourly to monthly
```%time dsr = ds.resample(freq='M', dim='time').load()```
on cheyenne takes:
```
CPU times: user 2min 24s, sys: 2min 11s, total: 4min 35s
Wall time: 3min 26s
```
compared to slightly longer time on my linux box:
```
CPU times: user 4min 12s, sys: 5min 44s, total: 9min 57s
Wall time: 4min 15s
```

This is using the sample dask.sh - two nodes, 36 cores each on cheyenne.

**Comment by mrocklin** [comment]
How many files are in your dataset?  @rabernat is this possibly the slow file system thing going on where we might want to avoid opening every individual file to check metadata?

**Comment by pwolfram** [comment]
Is this related to the 'too many open files' issue we had in xarray a while back?

**Comment by naomi-henderson** [comment]
In this test case, I open only 12 files in order to process a single year of a 55-year dataset.  Is there a way to turn off the metadata and coordinate checking entirely? Ryan and I did pre-process the files, trying to avoid of the overhead of checking metadata.

**Comment by rabernat** [comment]
> Is this related to the 'too many open files' issue we had in xarray a while back?

I doubt it. I think that would raise an error.

> is this possibly the slow file system thing going on where we might want to avoid opening every individual file to check metadata?

That is my hypothesis. We have already documented slow performance on `open_mfdataset` in pydata/xarray#1385. Xarray always eagerly reads all the dimension coordinates in each file during `open_mfdataset` because it needs them to align the different files. And note that these reads do _not_ get parallelized by dask; they happen in serial (at least as far as I understand). This could be fixed with a "fast path" option when one knows that the coordinates already aligned. But four minutes to open 12 files is extreme! We should dig deeper into the cause.

Fancy filesystems sometimes have caching that is invisible to the user. It's possible that caching is influencing the results. If the initial scan during `open_mfdataset` triggers the files to be cached, then the subsequent reads will be much faster.

@naomi-henderson, could you try runnning
```python
%prun  ds = xr.open_mfdataset(all_files, preprocess=set_correct_coords)
```
to get more detailed profiling?
That would give us more useful profiling information.

**Comment by mrocklin** [comment]
It looks like the data is already on Cheyenne, so at this point we can probably take over and profile ourselves as well.

**Comment by naomi-henderson** [comment]
Thanks @mrocklin , I will leave the rest to you, but I did try the profiling:
```%prun  ds = xr.open_mfdataset(all_files, preprocess=set_correct_coords)```

 very long, but here is the top:
```
         898484 function calls (877661 primitive calls) in 176.264 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     7078  172.504    0.024  172.504    0.024 {method 'acquire' of '_thread.lock' objects}
      114    2.147    0.019    2.156    0.019 {built-in method _operator.getitem}
       12    0.334    0.028    0.335    0.028 netCDF4_.py:182(_open_netcdf4_group)
       24    0.051    0.002    0.052    0.002 {netCDF4._netCDF4.num2date}
       75    0.050    0.001    0.050    0.001 {built-in method _thread.start_new_thread}
       24    0.038    0.002    0.042    0.002 conventions.py:222(nctime_to_nptime)
     7799    0.030    0.000    0.034    0.000 core.py:159(get_dependencies)
93089/92985    0.029    0.000    0.049    0.000 {built-in method builtins.isinstance}
      884    0.028    0.000    0.219    0.000 core.py:285(top)
5826/3426    0.024    0.000    0.032    0.000 indexing.py:363(shape)
     1768    0.024    0.000    0.037    0.000 inspect.py:2829(_bind)
      742    0.023    0.000    0.379    0.001 core.py:2462(elemwise)
     1432    0.021    0.000    0.022    0.000 local.py:300(release_data)
      884    0.021    0.000    0.311    0.000 core.py:2082(atop)
    10934    0.020    0.000    0.022    0.000 core.py:985(shape)
     1202    0.015    0.000    0.047    0.000 base.py:420(tokenize)
     7330    0.015    0.000    0.113    0.000 functoolz.py:281(__call__)
      890    0.014    0.000    0.052    0.000 core.py:231(broadcast_dimensions)
     3677    0.013    0.000    0.016    0.000 {built-in method builtins.sorted}
        3    0.012    0.004    0.012    0.004 {method 'read' of '_io.FileIO' objects}
     1947    0.012    0.000    0.017    0.000 itertoolz.py:66(groupby)
      166    0.012    0.000    0.022    0.000 optimize.py:456(fuse)
      166    0.012    0.000  172.712    1.040 local.py:405(get_async)
    18088    0.011    0.000    0.031    0.000 {built-in method builtins.getattr}
     1142    0.011    0.000    0.029    0.000 core.py:951(__new__)
     1267    0.010    0.000    0.061    0.000 local.py:489(fire_task)
      890    0.010    0.000    0.022    0.000 core.py:2016(unify_chunks)
     8713    0.010    0.000    0.018    0.000 abc.py:178(__instancecheck__)
     1267    0.010    0.000  172.524    0.136 queue.py:147(get)
10507/9716    0.010    0.000    0.175    0.000 {built-in method builtins.hasattr}
     1849    0.010    0.000    0.010    0.000 threading.py:215(__init__)
7522/4570    0.009    0.000    0.026    0.000 utils.py:410(__call__)
     4200    0.008    0.000    0.036    0.000 core.py:2406(is_scalar_for_elemwise)
     2534    0.008    0.000    0.014    0.000 threading.py:334(notify)
     8310    0.008    0.000    0.010    0.000 {method 'update' of 'dict' objects}
     8840    0.008    0.000    0.010    0.000 functoolz.py:136(__get__)
     1267    0.008    0.000    0.037    0.000 local.py:316(finish_task)
43057/41827    0.008    0.000    0.008    0.000 {built-in method builtins.len}
      884    0.007    0.000    0.018    0.000 functoolz.py:174(__init__)
```

**Comment by pwolfram** [comment]
I'm guessing if you try `%prun ds = xr.open_mfdataset(all_files, preprocess=set_correct_coords, autoclose=False)` you get the same results because `autoclose=False` by default: https://github.com/pydata/xarray/blob/27132fba1e465955e76ed155ba6cbb769c0904df/xarray/backends/api.py#L135, so this is something else as @rabernat notes.

**Comment by mrocklin** [comment]
So there is a ton of filesystem stuff going on (which is not surprising).  I'm also working through the notebook now with a cluster active and am seeing some odd behavior.  If anyone with knowledge of XArray internals has some time I would appreciate a video chat where I screenshare and we look at a few things.   There is a lot of short-sequential behavior going on.

cc @jhamman or @shoyer if he is interested in seeing things run on a cluster.

**Comment by mrocklin** [comment]
This line can take a surprisingly long time when we first start up

    all_files = glob(os.path.join(ddir, '**/*/latest/[h,u,v]*/*1960*.nc'), recursive=True)

**Comment by mrocklin** [comment]
The odd behavior that I'm seeing is tons of very small sequential computations.  It is as though somewhere in XArray there is a for loop calling compute many times over each of the chunks.

**Comment by rabernat** [comment]
Is that sequential computation happening during `open_mfdataset`?

**Comment by mrocklin** [comment]
Yes, also resample

**Comment by mrocklin** [comment]
If you'd like to see I can easily spin up a cluster and screenshare

**Comment by mrocklin** [comment]
Or maybe I'll make a quick screencast

**Comment by rabernat** [comment]
Neither of these operations are parallelized. They are basically loops.

open_mfdataset just [opens each file in a list](https://github.com/pydata/xarray/blob/master/xarray/backends/api.py#L538) and then [loops over them again](https://github.com/pydata/xarray/blob/master/xarray/core/combine.py#L249) when concatenating.

`resample` works like `groupby`: it loops over each group. The inability to parallelize groupby is a known issue (pydata/xarray#585) which may have just very recently been fixed by pydata/xarray#1517). (Not sure that PR gets all the way to groupby though.)

You have arrived one of the cruxes of xarray's performance bottlenecks. Welcome!

**Comment by rabernat** [comment]
> If you'd like to see I can easily spin up a cluster and screenshare

I personally don't need to see. I think I know exactly what your bokeh dashboard will show, because I see it myself every time I do these things: serial dispatching of dask tasks to the cluster.

A screencast would be very useful, however, to communicate the problem to a wider audience (e.g. xarray devs).

**Comment by mrocklin** [comment]
https://www.youtube.com/watch?v=nqud0JdaI6E&feature=youtu.be

**Comment by mrocklin** [comment]
@rabernat can you point me to that loop?

**Comment by rabernat** [comment]
> @rabernat can you point me to that loop?

You mean the one in `groupby`? I'm pretty sure the core loop is here:
https://github.com/pydata/xarray/blob/master/xarray/core/groupby.py#L609

**Comment by jhamman** [comment]
@mrocklin - I'm guessing you're working off the last release of xarray. I think https://github.com/pydata/xarray/pull/1551 and a few other associated improvements may help with some of what you're seeing here. These changes are on the master branch now and may be worth trying out.

**Comment by mrocklin** [comment]
Yes, that helped considerably.  I also made the chunk sizes along time to be smaller, which both helped a lot with the parallelism and avoided crashing the main machine

    ds = xr.open_mfdataset(all_files, chunks={'time': 10}, preprocess=set_correct_coords)

The computations at the end happen in around 20-30 seconds now.  I think that the next largest performance boost for them would be to enable persist so that we can keep data in RAM rather than recompute each time.

**Comment by mrocklin** [comment]
Yeah, most of the time in the plotting lines is just redoing the data loading from disk.  

Updated Notebook: https://gist.github.com/4800484676155745f3cf318bce2e78a4

**Comment by shoyer** [comment]
Currently, a call to `open_mfdataset` calls `open_dataset` many times in a loop, which means that coordinate data is loaded serially to create indexes. We could significantly speed this up by refactoring to either (a) load coordinate data in parallel or (b) skip loading it in cases where we know coordinate data should be consistent ahead of time (e.g., by using something like [NcML](https://www.unidata.ucar.edu/software/thredds/current/netcdf-java/ncml/Aggregation.html)).

**Comment by naomi-henderson** [comment]
Much better @mrocklin  !   The new version is also working much faster on my machine.  

Howver, I am having trouble with your new resample line:
```dsr = ds.resample(time='M').mean()```
which does not work on my linux box:
```---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<timed exec> in <module>()

TypeError: resample() got an unexpected keyword argument 'time'
```
instead I stilll need:
```ds.resample(freq='M', dim='time')```

Is this a difference in our versions of xarray? (0.9.6)

**Comment by mrocklin** [comment]
Yes, I think so.  I got this warning message when I switched to xarray master.  I suspect that if you also switch then things will get even faster for you as well.

    pip install git+https://github.com/pydata/xarray.git --upgrade

**Comment by rabernat** [comment]
Now that we have the sphinx documentation site working, this use-case notebook can be added into `docs/use_cases/`.

@naomi-henderson: would you be comfortable making a pull request to add the latest and greatest version of your notebook to this repo?

**Comment by naomi-henderson** [comment]
@rabernat:  I will give it a try, but it will have to wait until tomorrow.  I will let you know if I have any trouble with the sphinx.

**Comment by rabernat** [comment]
great! you don't necessarily have to build the docs yourself locally. You can just add your notebook in the `docs/use_cases/` folder and then add it to the toctree in `docs/index.rst`.
